clients/ts/signalr/spec/HttpConnection.spec.ts

```diff
@@ -287,6 +288,150 @@ describe("HttpConnection", () => {
+            skipNegotiation: true,
+            transport: HttpTransportType.LongPolling,
+        } as IHttpConnectionOptions;
+
+        const connection = new HttpConnection("http://tempuri.org", options);
+        try {
+            await connection.start(TransferFormat.Text);
+            fail();
+            done();
+        } catch (e) {
+            // WebSocket is created when the transport is connecting which happens after
+            // negotiate request would be sent. No better/easier way to test this.
+            expect(e.message).toBe("Negotiation can only be skipped when using the WebSocket transport directly.");
+            done();
+        }
+    });
+
+    it("redirects to url when negotiate returns it", async (done) => {
+        let firstNegotiate = true;
+        let firstPoll = true;
+        const httpClient = new TestHttpClient()
+            .on("POST", /negotiate$/, (r) => {
+                if (firstNegotiate) {
+                    firstNegotiate = false;
+                    return { url: "https://another.domain.url/chat" };
+                }
+                return {
+                    availableTransports: [{ transport: "LongPolling", transferFormats: ["Text"] }],
+                    connectionId: "0rge0d00-0040-0030-0r00-000q00r00e00",
+                };
+            })
+            .on("GET", (r) => {
+                if (firstPoll) {
+                    firstPoll = false;
+                    return "";
+                }
+                return new HttpResponse(204, "No Content", "");
+            });
+
+        const options: IHttpConnectionOptions = {
+            ...commonOptions,
+            httpClient,
+            transport: HttpTransportType.LongPolling,
+        } as IHttpConnectionOptions;
+
+        try {
+            const connection = new HttpConnection("http://tempuri.org", options);
+            await connection.start(TransferFormat.Text);
+        } catch (e) {
+            fail(e);
+            done();
+        }
+
+        expect(httpClient.sentRequests.length).toBe(4);
+        expect(httpClient.sentRequests[0].url).toBe("http://tempuri.org/negotiate");
+        expect(httpClient.sentRequests[1].url).toBe("https://another.domain.url/chat/negotiate");
+        expect(httpClient.sentRequests[2].url).toMatch(/^https:\/\/another\.domain\.url\/chat\?id=0rge0d00-0040-0030-0r00-000q00r00e00/i);
+        expect(httpClient.sentRequests[2].url).toMatch(/^https:\/\/another\.domain\.url\/chat\?id=0rge0d00-0040-0030-0r00-000q00r00e00/i);
+        done();
+    });
+
+    it("fails to start if negotiate redirects more than 100 times", async (done) => {
+        const httpClient = new TestHttpClient()
+            .on("POST", /negotiate$/, (r) => ({ url: "https://another.domain.url/chat" }));
+
+        const options: IHttpConnectionOptions = {
+            ...commonOptions,
+            httpClient,
+            transport: HttpTransportType.LongPolling,
+        } as IHttpConnectionOptions;
```

Comment:
nit: Is this `as IHttpConnectionOptions` required? I thought TS would see the shape of the JSON object and allow it automatically.